egs/librispeech/asr/simple_v1/tokenizer_ctc_att_transformer_decode.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python3
+
+# Copyright 2021 Xiaomi Corporation (Author: Guo Liyong)
+# Apache 2.0
+
+import argparse
+import logging
+import os
+import random
+import re
+import sys
+
+from pathlib import Path
+from typing import Union
+
+import k2
+import numpy as np
+import torch
+
+from kaldialign import edit_distance
+from lhotse import load_manifest
+from lhotse.dataset import K2SpeechRecognitionDataset, SingleCutSampler
+from lhotse.dataset.input_strategies import AudioSamples
+
+from espnet_utils.asr import ESPnetASRModel
+from espnet_utils.nnlm_evaluator import EspnetNNLMEvaluator
+from snowfall.common import store_transcripts
+from snowfall.common import write_error_stats
+from snowfall.decoding.lm_rescore import decode_with_lm_rescoring
+from snowfall.training.ctc_graph import build_ctc_topo
+
+
+def decode(dataloader: torch.utils.data.DataLoader,
+           model: None,
+           device: Union[str, torch.device],
+           ctc_topo: None,
+           G=None,
+           evaluator=None,
+           numericalizer=None,
+           num_paths=-1):
+    tot_num_cuts = len(dataloader.dataset.cuts)
+    num_cuts = 0
+    results = []
+    for batch_idx, batch in enumerate(dataloader):
+        assert isinstance(batch, dict), type(batch)
+        speech = batch['inputs'].squeeze()
+        lengths = batch['supervisions']['num_samples']
+        # Input as audio signal
+        if isinstance(speech, np.ndarray):
+            speech = torch.tensor(speech)
+
+        # data: (Nsamples,) -> (1, Nsamples)
+        speech = speech.unsqueeze(0)
+        speech = speech.to(torch.device(device))
+        lengths = lengths.to(torch.device(device))
+
+        nnet_output = model.encode(speech=speech, speech_lengths=lengths)
+        nnet_output = nnet_output.detach()
+
+        blank_bias = -1.0
+        nnet_output[:, :, 0] += blank_bias
+
+        supervision_segments = torch.tensor([[0, 0, nnet_output.shape[1]]],
+                                            dtype=torch.int32)
+
+        with torch.no_grad():
+            dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+
+            output_beam_size = 8
+            lattices = k2.intersect_dense_pruned(ctc_topo, dense_fsa_vec, 20.0,
+                                                 output_beam_size, 30, 10000)
+
+        use_whole_lattice = False
+        best_paths = decode_with_lm_rescoring(
+            lattices,
+            G,
+            evaluator,
+            num_paths=num_paths,
+            use_whole_lattice=use_whole_lattice)
+
+        token_int = list(
+            filter(
+                lambda x: x not in
+                [-1, 0, numericalizer.sos_idx, numericalizer.eos_idx],
+                best_paths.aux_labels.cpu().numpy()))
+
+        token = numericalizer.ids2tokens(token_int)
+
+        text = numericalizer.tokens2text(token)
+
+        ref = batch['supervisions']['text']
+        for i in range(len(ref)):
+            hyp_words = text.split(' ')
```

Comment:
What's the format of `text`?
Does `text` depend on `i`? If not, you can `split` it outside of the for loop.